Add webpack related theme development documentation (#100)
pandas-dev/pydata-bootstrap-sphinx-theme: README.md

to work on better configurability and extensibility. Feedback and contributions
are very welcome!

## Theme development

The theme is bundled via Webpack. In `./src/*` the theme related stylesheets and javascript is located. 2 entrypoints are available:

- Stylesheet: `./src/scss/theme.scss'
- Javascript: `./src/js/theme.js`

Both entrypoints will be bundled into `./pandas_sphinx_theme/static/{css,js}`.

Theme development was inspired by [ReadTheDocs](https://github.com/readthedocs/sphinx_rtd_theme).

### Steps to develop the theme:

> 1. Install yarn

Reviewer: Can you provide instructions for how to do each of these?

2. Install theme dependencies
3. Run development server
4. Build production assets

#### Install yarn

[Yarn](https://yarnpkg.com) is a package manager we use for managing Javascript and CSS dependencies.

Install via conda:

```bash

> conda install

Reviewer: `conda install yarn`?

Author: Indeed, well spotted.

```

Install alternative: https://classic.yarnpkg.com/en/docs/install#mac-stable.

#### Install theme dependencies

Install theme related dependencies:

```bash
yarn install
```

#### Run development server

```bash
yarn build:dev
```

> A development server is availlable at http://localhost:1919. When working on the theme, like editing stylesheets, javascript, .rst or .py files every saves reloads the development server. This reload includes bundling stylesheets, javascript and run sphinx again.

Reviewer: ```suggestion
A development server is available at http://localhost:1919. When working
on the theme, like editing stylesheets, javascript, .rst or .py files
every save reloads the development server. This reload includes bundling
stylesheets, javascript, and runing sphinx again.
```

Reviewer: What if running Sphinx is an expensive operation? E.g., the matplotlib docs take like 5 minutes to build from scratch if one were developing a theme...

Author: I think in theme development we are able to limit the extensive operations? In real documentation environments, the theme will be used, not developed. Am I correct on this one?

to work on better configurability and extensibility. Feedback and contributions
are very welcome!

## Theme development

The theme is bundled via Webpack. In `./src/*` the theme related stylesheets and javascript is located. 2 entrypoints are available:

> - Stylesheet: `./src/scss/theme.scss'

Reviewer: Would it make sense to factor this out into some SCSS modules and then stitch them together with a `master.scss` file that imports them? I've seen that as a common pattern for managing scss.

Author: I would surely make sense. You mean to rename `theme.scss` as the main entrypoint and split up styles in separate modules which are imported in `master.scss?`

Often the naming convention would be `index.scss` or `index.js` but maybe `master.*` is more clear?

Reviewer: Yep, that's what I was thinking. Index would be totally fine w me

- Javascript: `./src/js/theme.js`

Both entrypoints will be bundled into `./pandas_sphinx_theme/static/{css,js}`.

> Theme development was inspired by [ReadTheDocs](https://github.com/readthedocs/sphinx_rtd_theme).

Reviewer 2: ```suggestion
Theme development was inspired by the [ReadTheDocs sphinx theme](https://github.com/readthedocs/sphinx_rtd_theme).
```

### Steps to develop the theme:

1. Install yarn
2. Install theme dependencies
3. Run development server
4. Build production assets

#### Install yarn

[Yarn](https://yarnpkg.com) is a package manager we use for managing Javascript and CSS dependencies.

Install via conda:

```bash
conda install
```

> Install alternative: https://classic.yarnpkg.com/en/docs/install#mac-stable.

Reviewer 2: Maybe remove the mac specific anchor?

Author: Good point, missed it over pasting.

#### Install theme dependencies

Install theme related dependencies:

```bash
yarn install
```

#### Run development server

```bash
yarn build:dev
```

A development server is available at http://localhost:1919. When working
on the theme, like editing stylesheets, javascript, .rst or .py files
every save reloads the development server. This reload includes bundling

> stylesheets, javascript, and runing sphinx again.

Reviewer 2: ```suggestion
stylesheets, javascript, and running sphinx again.
```

> The follow files will be watched and reloaded on change:

Reviewer 2: ```suggestion
The following files will be watched and reloaded on change:
```

- ./src/js/theme.js
- ./src/scss/theme.scss
- ./docs/\*\*/\*.rst

> - ./docs/\*\*/\*.py

Reviewer 2: Are the py files inside pandas_sphinx_theme also whatched? 
And the html layouts?

Author: Currently not, but is is possible.

#### Build production assets

To build the new theme assets into the theme, run the following command.

```bash
yarn build:production
```

> 

Reviewer 2: Maybe mention that it is needed to run either build:dev or build:production before committing, as the generated files are included and tracked in the repo?

Author: Good suggestion, will add this.

Reviewer 2: I suppose that certainly the html layouts make sense to also watch?

to work on better configurability and extensibility. Feedback and contributions
are very welcome!

## Theme development

The theme is bundled via Webpack. In `./src/*` the theme related stylesheets and javascript is located. 2 entrypoints are available:

- Stylesheet: `./src/scss/index.scss'
- Javascript: `./src/js/index.js`

Both entrypoints will be bundled into `./pandas_sphinx_theme/static/{css,js}`.

Theme development was inspired by the [ReadTheDocs sphinx theme](https://github.com/readthedocs/sphinx_rtd_theme).

### Steps to develop the theme:

1. Install yarn
2. Install theme dependencies
3. Run development server
4. Build production assets

**Important:** in orde to commit changes to the theme, ensure you run `yarn build:production` so all assets will be bundled into `./pandas_sphinx_theme/static/`.

#### Install yarn

[Yarn](https://yarnpkg.com) is a package manager we use for managing Javascript and CSS dependencies.

Install via conda:

```bash
conda install yarn
```

Install alternative: https://classic.yarnpkg.com/en/docs/install.

#### Install theme dependencies

Install theme related dependencies:

```bash
yarn install

> ```

Reviewer 2: This you can run initially, but suppose you already did this, and at some point we update the dependencies (or you want to ensure you still have all up to date dependencies), what to do then? The same command?

Author: Indeed the same command. Yarn writes away its exact installed dependencies in the `yarn.lock` file. Even loosely defined dependencies won't mess up new installations because the lock file is your explicit reference.

Reviewer 2: Can you mention this explicitly that if the dependencies were updated, you need to run "yarn install" again?
